### DISCUSSION TITLE
classes/class_gpuparticles2d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#83] [User-contributed notes]
# classes/class_gpuparticles2d

Inherits: Node2D< CanvasItem< Node< Object A 2D particle emitter. Description: 2D particle node used to create a variety of particle systems and effects. GPUParticles2D features an emitter that gen...

https://docs.godotengine.org/en/stable/classes/class_gpuparticles2d.html

**Comment by FoxyFox909** [comment]
Friendly reminder to free your particles if you are instantiating them via code and only want one-shot or play-once-and-go-away particle Nodes.

You can simply connect its own `finished` signal to its own `queue_free` method, even after setting `emitting` to `true`, e.g. `new_particle_node.finished.connect(new_particle_node.queue_free)`

**Comment by MelPeslier** [comment]
How will you do to load an emmision mask from a viewport ?

**Reply by MelPeslier** [reply]
oups, sorry, questions not here

**Reply by Calinou** [reply]
You can use ViewportTexture for this, but you may run into a similar issue as decals and light projectors using ViewportTextures (in the sense that [the texture won't be updated in real-time out of the box](https://github.com/godotengine/godot/issues/73400)). If you run into this problem, you can force it to update every frame using the workaround listed in the issue I linked, but this can be slow with large textures.

**Reply by MelPeslier** [reply]
Thanks,

I found how to do it,
I capture an image, see for each pixel if it is visible, and then I transfer this data as an image to the particle shader converted to a shadermaterial !

**Comment by hrk4649** [comment]
To use emit_particles(), you need to change the project rendering method to Forward+.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/100369
